Parse/ParseClientConfiguration.m

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (c) 2015-present, Parse, LLC.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "ParseClientConfiguration.h"
+#import "ParseClientConfiguration_Private.h"
+
+#import "PFAssert.h"
+#import "PFApplication.h"
+#import "PFCommandRunningConstants.h"
+#import "PFFileManager.h"
+#import "PFHash.h"
+
+@implementation ParseClientConfiguration
+
+///--------------------------------------
+#pragma mark - Init
+///--------------------------------------
+
++ (instancetype)emptyConfiguration {
+    return [super new];
+}
+
+- (instancetype)initWithBlock:(void (^)(id<ParseMutableClientConfiguration>))configurationBlock {
+    self = [super init];
+    if (!self) return nil;
+
+    _networkRetryAttempts = PFCommandRunningDefaultMaxAttemptsCount;
+
+    configurationBlock(self);
+
+    return self;
+}
+
++ (instancetype)configurationWithBlock:(void (^)(id<ParseMutableClientConfiguration>))configurationBlock {
+    return [[self alloc] initWithBlock:configurationBlock];
+}
+
+///--------------------------------------
+#pragma mark - Properties
+///--------------------------------------
+
+- (void)setApplicationId:(NSString *)applicationId {
+    PFConsistencyAssert([applicationId length], @"'applicationId' should not be nil.");
+    _applicationId = [applicationId copy];
+}
+
+- (void)setClientKey:(NSString *)clientKey {
+    PFConsistencyAssert([clientKey length], @"'clientKey' should not be nil.");
+    _clientKey = [clientKey copy];
+}
+
+- (void)setApplicationGroupIdentifier:(NSString *)applicationGroupIdentifier {
+    PFConsistencyAssert(
```

Comment:
Do we want to let people specify `nil` group identifier?
I think this needs to be the other way around.